### ISSUE TITLE
SurfaceControl needs system apps permission

### ISSUE THREAD

**Issue by recnac-itna** [Genymobile/scrcpy#1083]
Instead of using `adb connect` to push and run "scrcpy-server.jar", I want to launch the scrcpy server in Android application directly. 

But I am blocked in accessing hidden method of `SurfaceControl`. [This post](https://stackoverflow.com/questions/39490725/invoking-createdisplay-on-android-view-surfacecontrol-using-reflection-returns-n) said I need system permission (android.permission.ACCESS_SURFACE_FLINGER), but "permission is only granted to system apps". 

So how can I launch the scrcpy server in my Android application without the help of adb? Or another way of thinking, can I replace `SurfaceControl` with other library which doesn't need system permission? 

Thanks for any help.

**Comment by rom1v** [comment]
That's the main reason why it is executed via `adb shell`.
https://blog.rom1v.com/2018/03/introducing-scrcpy/#run-a-java-main-on-android

From an app, you need to use the [`MediaProjection`](https://developer.android.com/reference/android/media/projection/MediaProjection) API.

**Comment by recnac-itna** [comment]
Yes, that is what I tried now. I have removed adb process and replaced `SurfaceControl` logic with `MediaProjection`, and it works without root.

The only drawback is that a pop-up box will pop up to let the user agree when use SystemService `MEDIA_PROJECTION_SERVICE`.

**Comment by rom1v** [comment]
> The only drawback is that a dialog will pop up to let the user agree

Yes, and also you cannot control (injecting events requires system permissions too).

**Comment by recnac-itna** [comment]
> Yes, and also you cannot control (injecting events requires system permissions too).

Sigh, you are right. It still needs system permissions. Do we have another way to remote control without system permissions ?
